### PR TITLE
Convert Table Potential to v3 API

### LINKED ISSUE #360: Port Table potentials to v3 API
HOOMD's tabulated potential implementation and API are being very widely used. Their implementation, and Python API in particular, could use a thorough review and be improved if possible.

### PR DESCRIPTION
**Pull request by danevans09**
## Description

This converts the Table potential to the v3 API. I created an `EvaluatorPairTable.h` file to follow the format of other pair potentials and changed it so the force and energy tables are directly set as numpy arrays for each pair of particle types.
## Motivation and context

The Table potential has not yet been converted to the v3 API.

Works toward #360 

## How has this been tested?

I added valid and invalid Table potential arguments to `test_potential.py`.

## Change log

```

```

## Checklist:

- [X] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [X] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [X] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/EvaluatorPairTable.h
Comment (danevans09) [inline hoomd/md/EvaluatorPairTable.h:76]: The old code used a `GlobalArray` to store the force and energy values but I wasn't sure how to access `exec_conf` to initialize a `GlobalArray`

File: hoomd/md/EvaluatorPairTable.h
Comment (danevans09) [inline hoomd/md/EvaluatorPairTable.h:183]: This should be: 

`(std::sqrt(rcutsq) - std::sqrt(ronsq)) / width` 

but `ronsq` is not in the constructor arguments. Should I add a `needsRon` method to every pair potential or change the constructor for `EvaluatorPairTable`?

Comment (joaander) [reply hoomd/md/EvaluatorPairTable.h:183]: The minimum r for the table potential and r_on used in the `PotentialPair` framework are two different things. You should add `r_min` to the `param_type` struct for the table potential and use that here.

**Review by joaander — COMMENTED** [review]
Dan, this is a great start! I was thinking we would just write a new interface to the old code - but your solution to re-implement as a `PotentialPair` evaluator is much better. It also solves some of the user requests we had (different widths for different interactions).

It is possible to use `ManagedArray` to manage the table values in this framework, but it will take some careful modification of the GPU kernels in `PotentialPairGPU` to get this working on the GPU. The CPU implementation should work without needing modification.

For now, you can update the code to use `ManagedArray` instead of `std::vector` and it should work on the CPU. `ManagedArray` doesn't get an `ExecutionConfiguration`, instead, it gets a boolean if it should use the GPU or not. You can just leave this as a default `false` for now. I need to look more closely at what changes are needed to `PotentialPair` to get the proper value passed down. I remember that took a while to get right in `IntegratorHPMCMono`.

File: hoomd/md/EvaluatorPairTable.h
Comment (b-butler) [inline hoomd/md/EvaluatorPairTable.h:133]: Why the `1 *`?

File: hoomd/md/module-md.cc
Comment (b-butler) [inline hoomd/md/module-md.cc:228]:
```suggestion
```

File: hoomd/md/pair/pair.py
Comment (b-butler) [inline hoomd/md/pair/pair.py:567]: Table doesn't support any of these though?

File: hoomd/md/pytest/test_potential.py
Comment (b-butler) [inline hoomd/md/pytest/test_potential.py:150]:
```suggestion
        if not (isinstance(valid_value, (list, np.ndarray)):
```

File: hoomd/md/pytest/test_potential.py
Comment (b-butler) [inline hoomd/md/pytest/test_potential.py:153]:
```suggestion
        if not (isinstance(valid_value, (str, np.ndarray)):
```

File: hoomd/md/EvaluatorPairTable.h
Comment (b-butler) [inline hoomd/md/EvaluatorPairTable.h:51]: This is not the case currently.

**Review by b-butler — COMMENTED** [review]
Overall, the PR looks good. I have a few comments on documentation, and on design potentially, but I want to see what @joaander  thinks of the PR before merging.

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:5]:
```suggestion
```
We no longer assign maintainers per file.

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:16]:
```suggestion
```
We no longer support doxygen documentation builds.

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:48]: I don't think this is accurate. If the number of bins is 10 and rmax is 1, then the potential is defined at the points 0.0, 0.1, ... 0.9. V and F are zero for r >= rmax.

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:45]: I think rmax is misleading in these comments. There is no rmax variable stored or computed in the new code. I would update the documentation to refer to r_cut instead of rmax.

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:82]: We aren't using AVX in the table potentials. You can allocate the arrays with size `width` and remove the `, 32` and all the alignment size computations.

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:88]: You could write this with less code using the C++ standard library copy algorithm: https://www.cplusplus.com/reference/algorithm/copy/

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:93]: This error check should occur first and it should check the lengths of the Python arrays the users provides, not the ones we allocate.

Also, why is "width" a separate parameter? You should be able to determine the width by inspecting the length of the arrays provided by the user in v["V"] and v["F"].

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:119]: @b-butler 

Are numpy arrays pickleable? Do we need to make changes here so that table potentials will work with the new pickling infrastructure?

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:179]: This code is used in the inner loop, let's not trust the compiler/optimizer. Only call sqrt once for each unique value and then store that in a variable and reuse it.

Also, use fast::sqrt instead of std::sqrt.

```
Scalar rcut = fast::sqrt(rcutsq);
Scalar r = fast::sqrt(rsq);
```

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:184]: Souldn't this be
`value_f = (r - rmin) / delta_r`
?

File: hoomd/md/EvaluatorPairTable.h
Comment (joaander) [inline hoomd/md/EvaluatorPairTable.h:192]: Handle out of bounds memory accesses:
```suggestion
                Scalar V0 = V_table[value_i];
                Scalar V1 = 0;
                if (value_i + 1 < width)
                    V1 = V_table[value_i + 1];
```

We might also be concerned about underflow or overflow of value_i for some unexpected floating point rounding conditions. We could write this more defensively by always computing `value_f` and `value_i` (as an `int`), then change the `if (rsq ...)` condition to `if (value_i >= 0 && value_i < width)`

File: hoomd/md/pair/__init__.py
Comment (joaander) [inline hoomd/md/pair/__init__.py:27]:
```suggestion
    Table,
    TWF
```
A comes before W.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks!

I reviewed the implementation, but not the documentation or tests. Change requests indicated.

File: hoomd/md/EvaluatorPairTable.h
Comment (danevans09) [inline hoomd/md/EvaluatorPairTable.h:204]: Should this 0.5 be removed? The tests that compare the simulation forces to forces calculated by differentiating the simulation energy are failing by a factor of 1/2

Comment (joaander) [reply hoomd/md/EvaluatorPairTable.h:204]: Yes, remove the 0.5. The code that calls the evaluator multiplies by 0.5 where necessary.

**Review by AlainKadar — APPROVED** [review]

File: hoomd/md/EvaluatorPairTable.h
Comment (b-butler) [reply hoomd/md/EvaluatorPairTable.h]: I think a NumPy array makes sense since that is how I would think to compute a table potential.

File: hoomd/md/EvaluatorPairTable.h
Comment (b-butler) [reply hoomd/md/EvaluatorPairTable.h:110]: NumPy arrays are pickable.

File: hoomd/md/pytest/test_potential.py
Comment (b-butler) [inline hoomd/md/pytest/test_potential.py:131]: The parenthesis around `isinstance` are unnecessary.
```suggestion
        if not isinstance(valid_value, (list, np.ndarray)):
```

File: hoomd/md/pytest/test_potential.py
Comment (b-butler) [inline hoomd/md/pytest/test_potential.py:134]:
```suggestion
        if not isinstance(valid_value, (str, np.ndarray)):
```

File: hoomd/md/pair/pair.py
Comment (b-butler) [inline hoomd/md/pair/pair.py:525]: Docstring needs to be formatted.

**Review by b-butler — CHANGES_REQUESTED** [review]
I have requested some changes that need to be made.

**Comment by b-butler** [comment]
@danevans09 sorry I hadn't finished making all the changes I was going to so I had compilation errors.

File: hoomd/md/EvaluatorPairTable.h
Comment (b-butler) [inline hoomd/md/EvaluatorPairTable.h:86]: C++ style casts are more descriptive of the expected type of cast and can catch more programming errors. It is also more clear for come complex expressions (which this is not admittedly).
```suggestion
                V_table = ManagedArray<Scalar>(static_cast<unsigned int>(width), managed, 32);
                F_table = ManagedArray<Scalar>(static_cast<unsigned int>(width), managed, 32);
```

File: hoomd/md/module-md.cc
Comment (b-butler) [inline hoomd/md/module-md.cc:378]: When this is working for the GPU we will need to export that to Python as well.

File: hoomd/md/PotentialPairGPU.cuh
Comment (b-butler) [inline hoomd/md/PotentialPairGPU.cuh:182]:
```suggestion
            reinterpret_cast<int*>(s_params)[cur_offset + threadIdx.x] = reinterpret_cast<int*>(d_params)[cur_offset + threadIdx.x];
```

File: hoomd/md/PotentialPairGPU.cuh
Comment (b-butler) [inline hoomd/md/PotentialPairGPU.cuh:188]:
```suggestion
    auto s_extra = reinterpret_cast<char*>(s_params + typpair_idx.getNumElements());
```

File: hoomd/md/PotentialPairGPU.cuh
Comment (b-butler) [inline hoomd/md/PotentialPairGPU.cuh:464]: Use static or reinterpret cast as necessary.

**Review by b-butler — COMMENTED** [review]
I know it is not done yet, but I went ahead and gave the changes a look and left some comments.

File: hoomd/md/module-md.cc
Comment (danevans09) [reply hoomd/md/module-md.cc:378]: I believe this is exporting the old `Table `potential implementation

Comment (b-butler) [reply hoomd/md/module-md.cc:378]: It is; I meant that we will need to expose the Table potential on the GPU too, similar to the other pair potentials.

File: hoomd/md/PotentialPairGPU.cuh
Comment (danevans09) [inline hoomd/md/PotentialPairGPU.cuh:462]: This line is causing the code to crash, however all of the tests in `test_potential.py` pass when I comment out this line

Comment (joaander) [reply hoomd/md/PotentialPairGPU.cuh:462]: OK, I will take a look.

Comment (joaander) [reply hoomd/md/PotentialPairGPU.cuh:462]: Look at the way HPMC handles this. It has a separate `allocate_shared` call that doesn't attempt to write to the pointer. I don't know why the aniso potential doesn't, but it should.

Comment (vyasr) [reply hoomd/md/PotentialPairGPU.cuh]: Did some quick digging here. I introduced the aniso code in #457, which was merged on Sept 16 2019. Jens introduce `allocate_shared` in #408, which was opened well before #457 but didn't get merged until about a month after. It looks like we just failed to synchronize those changes at the time since they appeared orthogonal.

Comment (joaander) [reply hoomd/md/PotentialPairGPU.cuh]: Looking further, the aniso potential is also using global arrays which can't hold nested data structures like managed arrays. I'm currently working to convert `PotentialPairGPU` to use `std::vector<param_type, managed_allocator<param_type> >` as HPMC does.

Comment (vyasr) [reply hoomd/md/PotentialPairGPU.cuh]: Due to the size and long-lived nature of a lot of Jens's PRs that made these changes it's not surprising that we ended up out of sync, as I recall you and I were sort of feeling our way through Jens's unified memory handling when we worked on adding shared memory support to anisotropic potentials. Hopefully the changes aren't terribly difficult to make.

**Review by joaander — COMMENTED** [review]
Thanks! This is ready for another round of review.

I still need to review the user-facing documentation.

**Review by joaander — APPROVED** [review]
Thanks! Subject to @b-butler approving the changes in `parameterdicts.py`, this looks good to me. 

I applied my comments directly as changes to the pair.Table documentation.

File: hoomd/md/pair/pair.py
Comment (b-butler) [inline hoomd/md/pair/pair.py:597]:
```suggestion
          * ``V`` ((*N*,) `numpy.ndarray` of `float`, **required**) -
            the tabulated energy values :math:`[\\mathrm{energy}]`.

          * ``F`` ((*N*,) `numpy.ndarray` of `float`, **required**) -
            the tabulated force values :math:`[\\mathrm{force}]`. Must have the
            same length as ``V``.
```
This is equivalent and the way NumPy documents and recommends using for methods/functions such as `astype`.

**Review by b-butler — COMMENTED** [review]
I have a few comments that I would like addressed, but I don't want to hold this up, so I will just leave it as a comment. Feel free to merge once resolved without another review by me.

**Review by b-butler — APPROVED** [review]

File: hoomd/md/pair/pair.py
Comment (joaander) [reply hoomd/md/pair/pair.py]: I realize I'm against the bandwagon here, but just because something is recommended in the docs doesn't mean it is a good practice. In this case the proliferation of `import numpy as np` provides a fluid representation that serves no purpose other than to make example code look cute. In contrast with what the linked documentation says, fluid representations make concepts more difficult to understand, not easier. If the authors truly intend for all users to use `np`, then they should have named the library `np`.

I personally find it more useful to `import numpy` and  copy and past method names and arguments that use the full name from the documentation (e.g. `numpy.testing.assert_allclose` documented here: https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html). The full names appear in the documentation more often than the short versions. As such, HOOMD has a mix of using numpy as `numpy` and `np` in different files.

**Pull request merged**
